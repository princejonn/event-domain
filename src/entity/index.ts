export * from "./Aggregate";
export * from "./Saga";
export * from "./View";
