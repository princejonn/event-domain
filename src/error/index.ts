export * from "./AggregateAlreadyCreatedError";
export * from "./AggregateDestroyedError";
export * from "./AggregateNotCreatedError";
export * from "./AggregateNotDestroyedError";
export * from "./CausationMissingEventsError";
export * from "./CommandSchemaValidationError";
export * from "./ConcurrencyError";
export * from "./DomainError";
export * from "./EventStreamRevisionError";
export * from "./HandlerNotRegisteredError";
export * from "./IllegalEntityChangeError";
export * from "./MessageTypeError";
export * from "./MongoDuplicateKeyError";
export * from "./MongoNotUpdatedError";
export * from "./SagaAlreadyCreatedError";
export * from "./SagaCausationError";
export * from "./SagaDestroyedError";
export * from "./SagaNotCreatedError";
export * from "./SagaNotInStoreError";
export * from "./SagaRevisionError";
export * from "./ViewAlreadyCreatedError";
export * from "./ViewCausationError";
export * from "./ViewDestroyedError";
export * from "./ViewNotCreatedError";
export * from "./ViewRevisionError";
