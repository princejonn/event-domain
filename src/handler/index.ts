export * from "./AggregateCommandHandler";
export * from "./AggregateEventHandler";
export * from "./SagaEventHandler";
export * from "./ViewEventHandler";
