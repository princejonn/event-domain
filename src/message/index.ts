export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
