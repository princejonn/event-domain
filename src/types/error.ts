import { LindormErrorOptions } from "@lindorm-io/errors";

export interface DomainErrorOptions extends LindormErrorOptions {
  permanent: boolean;
}
