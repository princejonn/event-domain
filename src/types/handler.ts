export interface HandlerIdentifier {
  name: string;
  context: string;
}

export interface HandlerConditions {
  created?: boolean;
  permanent?: boolean;
}

export interface HandlerIdentifierExpectingStructure {
  name?: string;
  context?: string;
}

export interface HandlerIdentifierMultipleContexts {
  name: string;
  context?: Array<string> | string;
}
