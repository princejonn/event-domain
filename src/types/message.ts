import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
