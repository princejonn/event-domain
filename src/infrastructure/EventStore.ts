import { Aggregate } from "../entity";
import { AggregateEventHandler } from "../handler";
import { CausationMissingEventsError } from "../error";
import { Command, DomainEvent } from "../message";
import { EventStoreInMemory } from "./in-memory";
import { EventStoreMongo } from "./mongo";
import { AggregateIdentifier, IEventStore, EventStoreOptions } from "../types";
import { Logger } from "@lindorm-io/winston";
import { StoreType } from "../enum";
import { CAUSATION_EVENTS_SCHEMA, SAVE_AGGREGATE_SCHEMA } from "../schema";
import { filter, take } from "lodash";
import { validateSchemaAsync } from "../util";

export class EventStore<State> {
  private readonly logger: Logger;
  private eventStore: IEventStore;

  constructor(logger: Logger, options?: EventStoreOptions) {
    this.logger = logger.createChildLogger(["EventStore"]);

    const type = options?.type || StoreType.MEMORY;

    switch (type) {
      case StoreType.CUSTOM:
        this.eventStore = options.custom;
        break;

      case StoreType.MEMORY:
        this.eventStore = new EventStoreInMemory();
        break;

      case StoreType.MONGO:
        this.eventStore = new EventStoreMongo(this.logger, options.mongo);
        break;

      default:
        throw new Error("Invalid Store Type");
    }
  }

  public async save(
    aggregate: Aggregate<State>,
    causation: Command,
  ): Promise<Array<DomainEvent>> {
    this.logger.debug("save aggregate events to store", {
      aggregate: {
        destroyed: aggregate.destroyed,
        events: aggregate.events,
        loaded: aggregate.numberOfLoadedEvents,
        state: aggregate.state,
      },
      causation,
    });

    await validateSchemaAsync(SAVE_AGGREGATE_SCHEMA, {
      aggregate,
      causation,
      events: aggregate.events,
    });

    const causationEvents = filter(
      aggregate.events,
      (event) => event.causationId === causation.id,
    );
    const expectedEvents = take(aggregate.events, aggregate.numberOfLoadedEvents);

    await validateSchemaAsync(CAUSATION_EVENTS_SCHEMA, { causationEvents });

    if (causationEvents.length === 0) {
      throw new CausationMissingEventsError();
    }

    return await this.eventStore.save(causationEvents, expectedEvents);
  }

  public async load(
    aggregateIdentifier: AggregateIdentifier,
    eventHandlers: Array<AggregateEventHandler<State>>,
  ): Promise<Aggregate<State>> {
    this.logger.debug("load aggregate events from store", {
      aggregateIdentifier,
      eventHandlers,
    });

    const events = await this.eventStore.load(aggregateIdentifier);
    const aggregate = new Aggregate(
      {
        id: aggregateIdentifier.id,
        name: aggregateIdentifier.name,
        context: aggregateIdentifier.context,
        eventHandlers: eventHandlers,
      },
      this.logger,
    );

    for (const event of events) {
      await aggregate.load(event);
    }

    return aggregate;
  }
}
