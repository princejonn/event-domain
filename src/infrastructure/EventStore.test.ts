import MockDate from "mockdate";
import { Aggregate } from "../entity";
import { AggregateEventHandler } from "../handler";
import { CausationMissingEventsError, EventStreamRevisionError } from "../error";
import { Command, DomainEvent } from "../message";
import { EventStore } from "./EventStore";
import {
  aggregateIdentifier,
  commandWithData,
  eventHandlerWithData,
  logger,
} from "../test";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("EventStore", () => {
  describe("when creating a new event store with default options", () => {
    it("should succeed", () => {
      expect(new EventStore(logger)).toStrictEqual(expect.any(EventStore));
    });
  });

  describe("when creating a new event store with erroneous options", () => {
    test("should throw", () => {
      expect(
        () =>
          new EventStore(logger, {
            // @ts-ignore
            type: "wrong",
          }),
      ).toThrow();
    });
  });

  describe("when saving events to store", () => {
    let eventStore: EventStore<any>;
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      eventStore = new EventStore(logger);

      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [eventHandlerWithData],
        },
        logger,
      );

      await aggregate.apply(commandWithData, "eventWithData", commandWithData.data);
    });

    it("should resolve the appended events", async () => {
      await expect(
        eventStore.save(aggregate, commandWithData),
      ).resolves.toMatchSnapshot();
    });

    it("should save the causation events to the store", () => {
      // @ts-ignore
      expect(eventStore.eventStore.eventsByAggregate).toMatchSnapshot();
    });
  });

  describe("when loading events from store", () => {
    let eventStore: EventStore<any>;
    let handler1: AggregateEventHandler<any>;

    beforeAll(async () => {
      eventStore = new EventStore(logger);

      const command1 = new Command({
        id: "86e9a7ee-7455-4438-9c61-7e41bd0191d5",
        name: "command1",
        aggregate: aggregateIdentifier,
      });

      const event1 = new DomainEvent(
        {
          id: "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
          name: "event1",
          aggregate: aggregateIdentifier,
          data: { eventData: true },
        },
        command1,
      );

      handler1 = new AggregateEventHandler({
        eventName: "event1",
        aggregate: aggregateIdentifier,
        handler: async (ctx) => {
          ctx.mergeState(ctx.event.data);
        },
      });

      const aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [handler1],
        },
        logger,
      );

      // @ts-ignore
      await aggregate.handleEvent(event1);

      await eventStore.save(aggregate, command1);
    });

    it("should load all events for aggregate", async () => {
      await expect(
        eventStore.load(aggregateIdentifier, [handler1]),
      ).resolves.toMatchSnapshot();
    });
  });

  describe("when there are no events on causation", () => {
    let eventStore: EventStore<any>;
    let aggregate: Aggregate<any>;
    let command2: Command;

    beforeAll(async () => {
      eventStore = new EventStore(logger);

      const command1 = new Command({
        id: "86e9a7ee-7455-4438-9c61-7e41bd0191d5",
        name: "command1",
        aggregate: aggregateIdentifier,
      });

      command2 = new Command({
        id: "ce9e080c-ced9-4e26-9eb5-c6620b9c72e7",
        name: "command2",
        aggregate: aggregateIdentifier,
      });

      const event1 = new DomainEvent(
        {
          id: "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
          name: "event1",
          aggregate: aggregateIdentifier,
          data: { eventData: true },
        },
        command1,
      );

      const handler1 = new AggregateEventHandler({
        eventName: "event1",
        aggregate: aggregateIdentifier,
        handler: async (ctx) => {
          ctx.mergeState(ctx.event.data);
        },
      });

      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [handler1],
        },
        logger,
      );

      await aggregate.load(event1);
    });

    it("should throw", async () => {
      await expect(eventStore.save(aggregate, command2)).rejects.toThrow(
        CausationMissingEventsError,
      );
    });
  });

  describe("when event tails are different", () => {
    let eventStore: EventStore<any>;
    let aggregate: Aggregate<any>;
    let command2: Command;

    beforeAll(async () => {
      eventStore = new EventStore(logger);

      const command1 = new Command({
        id: "86e9a7ee-7455-4438-9c61-7e41bd0191d5",
        name: "command1",
        aggregate: aggregateIdentifier,
      });

      const event1 = new DomainEvent(
        {
          id: "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
          name: "event1",
          aggregate: aggregateIdentifier,
          data: { eventData: true },
        },
        command1,
      );

      const handler1 = new AggregateEventHandler({
        eventName: "event1",
        aggregate: aggregateIdentifier,
        handler: async (ctx) => {
          ctx.mergeState(ctx.event.data);
        },
      });

      command2 = new Command({
        id: "bfc357a9-d0ac-4429-8cc3-ef75ed9caa73",
        name: "command2",
        aggregate: aggregateIdentifier,
      });

      const event2 = new DomainEvent(
        {
          id: "6901ee53-e05d-412a-8221-890b1a4a4673",
          name: "event2",
          aggregate: aggregateIdentifier,
          data: { eventData: true },
        },
        command2,
      );

      const handler2 = new AggregateEventHandler({
        eventName: "event2",
        aggregate: aggregateIdentifier,
        handler: async (ctx) => {
          ctx.mergeState(ctx.event.data);
        },
      });

      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [handler1, handler2],
        },
        logger,
      );

      // @ts-ignore
      await aggregate.handleEvent(event1);

      await eventStore.save(aggregate, command1);

      // @ts-ignore
      await aggregate.handleEvent(event2);
    });

    it("should throw when the tails are different", async () => {
      await expect(eventStore.save(aggregate, command2)).rejects.toThrow(
        EventStreamRevisionError,
      );
    });
  });
});
