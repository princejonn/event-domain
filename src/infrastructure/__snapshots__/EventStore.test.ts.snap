// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventStore when loading events from store should load all events for aggregate 1`] = `
Object {
  "context": "defaultContext",
  "destroyed": false,
  "events": Array [
    DomainEvent {
      "aggregate": Object {
        "context": "defaultContext",
        "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
        "name": "aggregateName",
      },
      "causationId": "86e9a7ee-7455-4438-9c61-7e41bd0191d5",
      "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "data": Object {
        "eventData": true,
      },
      "delay": 0,
      "id": "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
      "mandatory": false,
      "name": "event1",
      "timestamp": 2022-01-01T08:00:00.000Z,
      "type": "event",
    },
  ],
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "name": "aggregateName",
  "numberOfLoadedEvents": 1,
  "state": Object {
    "eventData": true,
  },
}
`;

exports[`EventStore when saving events to store should resolve the appended events 1`] = `
Array [
  DomainEvent {
    "aggregate": Object {
      "context": "defaultContext",
      "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
      "name": "aggregateName",
    },
    "causationId": "6c2e5bbd-6643-469d-92ba-f6a140d42cb4",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {
      "name": "hello",
    },
    "delay": 0,
    "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "mandatory": false,
    "name": "eventWithData",
    "timestamp": 2022-01-01T08:00:00.000Z,
    "type": "event",
  },
]
`;

exports[`EventStore when saving events to store should save the causation events to the store 1`] = `
Object {
  "defaultContext.aggregateName.2a553a74-0f84-4a7d-b91e-2902666db98c": Array [
    DomainEvent {
      "aggregate": Object {
        "context": "defaultContext",
        "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
        "name": "aggregateName",
      },
      "causationId": "6c2e5bbd-6643-469d-92ba-f6a140d42cb4",
      "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "data": Object {
        "name": "hello",
      },
      "delay": 0,
      "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "mandatory": false,
      "name": "eventWithData",
      "timestamp": 2022-01-01T08:00:00.000Z,
      "type": "event",
    },
  ],
}
`;
